File: backend/api-compra/handler/swagger-handler.py
import os
import base64
import mimetypes
import json
import logging
import boto3

logger = logging.getLogger()
logger.setLevel(logging.INFO)

def lambda_handler(event, context):
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type, Authorization',
        'Access-Control-Allow-Methods': 'GET, OPTIONS'
    }

    # 1. Manejar preflight OPTIONS
    if event.get('httpMethod') == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({'message': 'Preflight OK'})
        }

    # 2. Servir Swagger UI sin verificación de token
    path = event.get('path', '')
    if '/docs/swagger' in path:
        try:
            base_path = os.path.join(os.path.dirname(__file__), '../docs/swagger-ui')
            proxy = event.get('pathParameters', {}).get('proxy', '')
            
            if '..' in proxy:
                return {
                    'statusCode': 403,
                    'headers': headers,
                    'body': json.dumps({'error': 'Path traversal no permitido'})
                }

            file_path = os.path.join(base_path, proxy) if proxy else os.path.join(base_path, 'index.html')

            with open(file_path, 'rb') as file:
                content = file.read()
            
            mime_type, _ = mimetypes.guess_type(file_path)
            content_type = mime_type if mime_type else 'application/octet-stream'

            return {
                'statusCode': 200,
                'headers': {
                    **headers,
                    'Content-Type': content_type
                },
                'body': base64.b64encode(content).decode('utf-8'),
                'isBase64Encoded': True
            }

        except FileNotFoundError:
            logger.error(f"Archivo no encontrado: {file_path}")
            return {
                'statusCode': 404,
                'headers': headers,
                'body': json.dumps({'error': 'Archivo no encontrado'})
            }
        except Exception as e:
            logger.error(f"Error interno: {str(e)}")
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps({'error': 'Error interno del servidor'})
            }

    # 3. Para otras rutas: Verificación de token
    token = event.get('headers', {}).get('Authorization') or event.get('headers', {}).get('authorization', '')
    if not token or not token.startswith('Bearer '):
        return {
            'statusCode': 401,
            'headers': headers,
            'body': json.dumps({'error': 'Token no proporcionado'})
        }

    try:
        lambda_client = boto3.client('lambda')
        response = lambda_client.invoke(
            FunctionName=os.environ['VALIDAR_TOKEN_FUNCTION_NAME'],
            InvocationType='RequestResponse',
            Payload=json.dumps({"token": token.split(' ')[1]})
        )
        validation = json.loads(response['Payload'].read())

        if validation['statusCode'] != 200:
            return {
                'statusCode': 403,
                'headers': headers,
                'body': json.dumps({'error': 'Token inválido'})
            }

        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({'message': 'Acceso autorizado'})
        }

    except Exception as e:
        logger.error(f"Error al validar token: {str(e)}")
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'error': 'Error al validar token'})
        }

File: backend/api-producto/handler/swagger-handler.js
const fs = require('fs');
const path = require('path');
const AWS = require('aws-sdk');
const lambda = new AWS.Lambda();

const mimeTypes = {
  '.html': 'text/html',
  '.js': 'text/javascript',
  '.css': 'text/css',
  '.json': 'application/json',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon'
};

exports.lambda_handler = async (event) => {
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': '*',
    'Access-Control-Allow-Methods': 'GET, OPTIONS'
  };

  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({ message: 'Preflight OK' })
    };
  }

  const rawAuth = event.headers.Authorization || event.headers.authorization || '';
  const token = rawAuth.startsWith('Bearer ') ? rawAuth.split(' ')[1] : rawAuth;

  if (!token) {
    return {
      statusCode: 401,
      headers,
      body: JSON.stringify({ error: 'Token no proporcionado' })
    };
  }

  try {
    const tokenResult = await lambda.invoke({
      FunctionName: process.env.VALIDAR_TOKEN_FUNCTION_NAME,
      InvocationType: 'RequestResponse',
      Payload: JSON.stringify({ token })
    }).promise();

    const validation = JSON.parse(tokenResult.Payload);
    if (validation.statusCode !== 200) {
      return {
        statusCode: 403,
        headers,
        body: JSON.stringify({ error: 'Token inválido o expirado' })
      };
    }

    const basePath = path.join(__dirname, '../docs/swagger-ui');
    const proxy = event.pathParameters?.proxy || '';
    
    if (proxy.includes('..')) {
      return {
        statusCode: 403,
        headers,
        body: JSON.stringify({ error: 'Path traversal no permitido' })
      };
    }

    const filePath = proxy 
      ? path.join(basePath, proxy)
      : path.join(basePath, 'index.html');

    const fileContent = fs.readFileSync(filePath);
    const ext = path.extname(filePath);
    const contentType = mimeTypes[ext] || 'application/octet-stream';

    return {
      statusCode: 200,
      headers: {
        ...headers,
        'Content-Type': contentType
      },
      body: fileContent.toString('base64'),
      isBase64Encoded: true
    };

  } catch (error) {
    console.error('Error:', error);
    if (error.code === 'ENOENT') {
      return {
        statusCode: 404,
        headers,
        body: JSON.stringify({ error: 'Archivo no encontrado' })
      };
    }
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ error: 'Error interno del servidor' })
    };
  }
};